internal/sink/namespaces.go

```diff
@@ -0,0 +1,116 @@
+package sink
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/oliveagle/jsonpath"
+
+	"github.com/openmeterio/openmeter/internal/ingest/kafkaingest/serializer"
+	"github.com/openmeterio/openmeter/pkg/models"
+)
+
+type MeterStore struct {
+	Meters []*models.Meter
+}
+
+type NamespaceStore struct {
+	namespaces map[string]*MeterStore
+}
+
+func NewNamespaceStore() *NamespaceStore {
+	return &NamespaceStore{
+		namespaces: make(map[string]*MeterStore),
+	}
+}
+
+func (n *NamespaceStore) AddMeter(namespace string, meter *models.Meter) {
+	if n.namespaces[meter.Namespace] == nil {
+		n.namespaces[meter.Namespace] = &MeterStore{
+			Meters: []*models.Meter{meter},
+		}
+	} else {
+		n.namespaces[meter.Namespace].Meters = append(n.namespaces[meter.Namespace].Meters, meter)
+	}
+}
+
+// validateEvent validates a single event by matching it with the corresponding meter if any
+func (a *NamespaceStore) validateEvent(ctx context.Context, event serializer.CloudEventsKafkaPayload, namespace string) error {
+	namespaceStore := a.namespaces[namespace]
+	if namespaceStore == nil {
+		// We drop events from unknown org
+		return NewProcessingError(fmt.Sprintf("namespace not found: %s", namespace), DROP)
+	}
+
+	// Validate a single event against multiple meters
+	found := 0
+	for _, meter := range namespaceStore.Meters {
+		if meter.EventType == event.Type {
+			found++
+			err := validateEventWithMeter(meter, event)
+			if err != nil {
+				return err
+			}
+			// A single event can match multiple meters so we cannot break the loop early
+		}
+	}
+
+	if found == 0 {
+		// Send to dead letter queue so we can show it to the user
+		return NewProcessingError(fmt.Sprintf("no meter found for event type: %s", event.Type), DEADLETTER)
+	}
+
+	return nil
+}
+
+// validateEventWithMeter validates a single event against a single meter
+func validateEventWithMeter(meter *models.Meter, ev serializer.CloudEventsKafkaPayload) *ProcessingError {
+	// We can skip count events with no group bys
+	if meter.Aggregation != models.MeterAggregationCount && len(meter.GroupBy) == 0 {
+		return nil
+	}
```

Comment:
Here is an idea to make this more readable:

- Validate group bys first (if there are no group bys, the loop will do nothing)
- if the meter is a count meter, exit with nil
- continue running the rest of the function without any further checks for aggregation

From what I can tell even the json unmarshal is unnecessary for the count aggregation, so reorganizing the function like this wouldn't just optimize it, but would also improve readability (unless there is a reason why this wouldn't work)